File: packages/sdk/src/polling.ts
/**
 * Polling utility for waiting on sandbox status transitions.
 */

import { SandboxError, TimeoutError } from "./errors.js"
import { request } from "./http.js"
import type {
  ApiSandboxResponse,
  SandboxInfo,
  SandboxStatus,
  SandboxWaitOptions,
} from "./types.js"
import { toSandboxInfo } from "./types.js"

/**
 * Poll GET /sandboxes/{id} until status matches `target`.
 *
 * Bails early if the sandbox reaches a terminal status (`failed` or `deleted`)
 * that isn't the target.
 *
 * @throws TimeoutError if the deadline elapses.
 * @throws SandboxError if the sandbox reaches a terminal failed/deleted state.
 */
export async function waitForStatus(
  sandboxId: string,
  target: SandboxStatus,
  baseUrl: string,
  apiKey: string,
  opts: SandboxWaitOptions = {},
): Promise<SandboxInfo> {
  const { timeoutMs = 60_000, intervalMs = 1_000, signal } = opts
  const deadline = Date.now() + timeoutMs
  let lastStatus: string | undefined

  while (Date.now() < deadline) {
    if (signal?.aborted) {
      throw new SandboxError("Polling aborted")
    }

    const raw = await request<ApiSandboxResponse>({
      method: "GET",
      url: `${baseUrl}/sandboxes/${sandboxId}`,
      headers: { "X-API-Key": apiKey },
      signal,
    })
    lastStatus = raw.status
    if (raw.status === target) {
      return toSandboxInfo(raw)
    }
    // Fail fast on terminal states
    if (target !== "failed" && raw.status === "failed") {
      throw new SandboxError(
        `Sandbox ${sandboxId} reached "failed" state before "${target}"`,
      )
    }
    if (target !== "deleted" && raw.status === "deleted") {
      throw new SandboxError(
        `Sandbox ${sandboxId} was deleted while waiting for "${target}"`,
      )
    }

    // Linear poll with ±20% jitter
    const jitter = intervalMs * (0.8 + Math.random() * 0.4)
    await sleep(jitter)
  }

  throw new TimeoutError(
    `Timed out after ${timeoutMs}ms waiting for sandbox ${sandboxId} ` +
      `to reach "${target}". Last status: "${lastStatus ?? "unknown"}".`,
  )
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms))
}

File: packages/python-sdk/src/superserve/_polling.py
"""Polling utility for waiting on sandbox status transitions."""

from __future__ import annotations

import asyncio
import random
import time
from typing import Optional

from ._config import ResolvedConfig
from ._http import api_request, async_api_request
from .errors import SandboxError, SandboxTimeoutError
from .types import SandboxInfo, SandboxStatus, to_sandbox_info


def wait_for_status(
    sandbox_id: str,
    target: SandboxStatus,
    config: ResolvedConfig,
    *,
    timeout_seconds: float = 60.0,
    interval_seconds: float = 1.0,
) -> SandboxInfo:
    """Poll GET /sandboxes/{id} until status matches target."""
    deadline = time.monotonic() + timeout_seconds
    last_status: Optional[str] = None

    while time.monotonic() < deadline:
        raw = api_request(
            "GET",
            f"{config.base_url}/sandboxes/{sandbox_id}",
            headers={"X-API-Key": config.api_key},
        )
        last_status = raw.get("status")
        if last_status == target.value:
            return to_sandbox_info(raw)

        # Fail fast on terminal states
        if target.value != "failed" and last_status == "failed":
            raise SandboxError(
                f"Sandbox {sandbox_id} reached 'failed' state before '{target.value}'"
            )
        if target.value != "deleted" and last_status == "deleted":
            raise SandboxError(
                f"Sandbox {sandbox_id} was deleted while waiting for '{target.value}'"
            )

        # Linear poll with ±20% jitter
        jitter = interval_seconds * (0.8 + random.random() * 0.4)
        time.sleep(jitter)

    raise SandboxTimeoutError(
        f"Timed out after {timeout_seconds}s waiting for sandbox {sandbox_id} "
        f'to reach "{target.value}". Last status: "{last_status or "unknown"}".'
    )


async def async_wait_for_status(
    sandbox_id: str,
    target: SandboxStatus,
    config: ResolvedConfig,
    *,
    timeout_seconds: float = 60.0,
    interval_seconds: float = 1.0,
) -> SandboxInfo:
    """Async variant of wait_for_status."""
    deadline = time.monotonic() + timeout_seconds
    last_status: Optional[str] = None

    while time.monotonic() < deadline:
        raw = await async_api_request(
            "GET",
            f"{config.base_url}/sandboxes/{sandbox_id}",
            headers={"X-API-Key": config.api_key},
        )
        last_status = raw.get("status")
        if last_status == target.value:
            return to_sandbox_info(raw)

        if target.value != "failed" and last_status == "failed":
            raise SandboxError(
                f"Sandbox {sandbox_id} reached 'failed' state before '{target.value}'"
            )
        if target.value != "deleted" and last_status == "deleted":
            raise SandboxError(
                f"Sandbox {sandbox_id} was deleted while waiting for '{target.value}'"
            )

        jitter = interval_seconds * (0.8 + random.random() * 0.4)
        await asyncio.sleep(jitter)

    raise SandboxTimeoutError(
        f"Timed out after {timeout_seconds}s waiting for sandbox {sandbox_id} "
        f'to reach "{target.value}". Last status: "{last_status or "unknown"}".'
    )
